thunder/dynamo/splitter.py

```diff
@@ -140,20 +144,40 @@ def is_thunder_supported_partition(node: torch.fx.Node) -> bool:
     # Call compile on the split region/s.
     thunder_compiled_fns = []
     submodule_to_compiled_fns = {}
+
+    # If the split_gm contains Thunder-traceable checkpoint operators, the checkpointed submodules are replaced, so we need to record the original split_gm
+    origin_split_gm = split_gm
+    if has_thunder_tracable_checkpoint:
+        import copy
+
+        split_gm = copy.deepcopy(split_gm)
+
     for node in split_gm.graph.nodes:
         if is_thunder_supported_partition(node):
             graph_module = getattr(split_gm, node.name)
+            # Replace the torch operators within the function called by activation checkpoint with the corresponding Thunder symbols
+            checkpoint_converter(split_gm, graph_module)
             jit_fn = thunder_jit(graph_module)
             # Update the node name from "submod_*" to "thunder_*" for more user-friendly names
             update_node_and_submodule(split_gm, node, node.name.replace("submod", "thunder"), jit_fn)
             thunder_compiled_fns.append(jit_fn)
-            submodule_to_compiled_fns[graph_module] = CompiledFunction(jit_fn, CompilerType.THUNDER)
+            if has_thunder_tracable_checkpoint:
+                submodule_to_compiled_fns[getattr(origin_split_gm, node.name.replace("thunder", "submod"))] = (
+                    CompiledFunction(jit_fn, CompilerType.THUNDER)
```

Comment:
Here the assumption is user will try to index with graph from `original_split_gm`. I assume it will fail with KeyError if graph came from `split_graph_module`.

I was thinking adding `original_split_gm` to `SubgraphInfo` could be confusing as to whether we should look at `split_graph_module` or `original_split_gm`. So maybe we should keep only one and add a boolean flag to denote if checkpointing occured. Wdyt?

Reply:
> Do we want to save the original split graph or the split graph after compile or both? Maybe we keep them both?
(for the graph by graph benchmarking we need the original module to compile for different backend)

So IIUC, we would need the `original_split_graph_module` for benchmarking utility. In that case, do you think we should rename `original_split_graph_module` -> `split_graph_module` (this will always be present in SubgraphInfo and be output of `fx.split_module`) and current `split_graph_module` -> `compiled_split_graph_module` (this will have thunder/inductor modules patched into the `split_graph_module`). 

With this the SubgraphInfo will look a bit like

```python
original_graph_module: torch.fx.GraphModule
split_graph_module: torch.fx.GraphModule
compiled_split_graph_module: torch.fx.GraphModule
thunder_compiled_fns: list[Callable] | None
submodule_to_compiled_functions: dict[torch.fx.GraphModule, CompiledFunction]
split_reasons: list | None = None
```

I think currently, my main confusion was while debugging which one of `original_split_graph_module` and `split_graph_module` should be referred. As both maybe same or different depending on whether `checkpointing` was present. (And one has to look at either graphs to determine whether it was present or not).

Reply:
Thank you @kshitij12345 , I'll get the `split_graph_module` as the original split module and the `compiled_split_graph_module` as the compiled one (with jitted func and converter changes if there are any), then in this way the naming is consistent w/ or w/o checkpointing.